plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilter.java

```diff
@@ -137,4 +142,106 @@ public void failed(Exception e) {
     public int order() {
         return Integer.MAX_VALUE;
     }
+
+    final static class BulkProcessingState implements Iterator<ActionRequest> {
```

Comment:
nit: looking for a better name for this. it does hold the state but it's also an iterator over requests. This makes me think that this class may be doing too much at the same time. My personal opinion though, just because this iterator does much more than a simple iterator. Maybe we should have an iterator and a state object separated?